make `@embedFile` support module-mapped names the same way as `@import`

Concrete example:

#### bootloader.zig

```zig
pub export fn _start() void {}
```

#### main.zig

```zig
const std = @import("std");
const blah = @embedFile("bootloader.elf");

pub fn main() !void {
    std.debug.print("first 3 bytes: " ++ blah[0..3] ++ "\n", .{});
}
```

#### build.zig

```zig
const std = @import("std");

pub fn build(b: *std.Build) void {
    const target = b.standardTargetOptions(.{});
    const optimize = b.standardOptimizeOption(.{});

    const bootloader = b.addExecutable(.{
        .name = "bootloader",
        .root_source_file = .{ .path = "bootloader.zig" },
        .target = .{
            .cpu_arch = .x86,
            .os_tag = .freestanding,
        },
        .optimize = .ReleaseSmall,
    });

    const exe = b.addExecutable(.{
        .name = "embed-generated-file",
        .root_source_file = .{ .path = "main.zig" },
        .target = target,
        .optimize = optimize,
    });
    exe.addAnonymousModule("bootloader.elf", .{
        .source_file = bootloader.getOutputSource(),
    });
    exe.install();
}
```

-----

This fully works according to the build system, but compilation of main.zig fails like this:

```
andy@ark ~/m/embed-generated-file> zig build
main.zig:2:25: error: unable to open 'bootloader.elf': FileNotFound
const blah = @embedFile("bootloader.elf");
                        ^~~~~~~~~~~~~~~~
referenced by:
    main: main.zig:5:42
    callMain: /home/andy/Downloads/zig/lib/std/start.zig:616:32
    remaining reference traces hidden; use '-freference-trace' to see all reference traces

error: embed-generated-file...
error: The following command exited with error code 1:
/home/andy/Downloads/zig/build-release/stage3/bin/zig build-exe /home/andy/misc/embed-generated-file/main.zig --cache-dir /home/andy/misc/embed-generated-file/zig-cache --global-cache-dir /home/andy/.cache/zig --name embed-generated-file --pkg-begin bootloader.elf /home/andy/misc/embed-generated-file/zig-cache/o/4fbab5d6b4239cb2e694a118ffff2202/bootloader --pkg-end --enable-cache 
```

You can see that it did, indeed, pass these flags:
```
--pkg-begin
bootloader.elf
/home/andy/misc/embed-generated-file/zig-cache/o/4fbab5d6b4239cb2e694a118ffff2202/bootloader
--pkg-end
```

This issue is to make zig honor those flags, and thereby allow `@embedFile("bootloader.elf")` to work.

--- a/src/Module.zig
+++ b/src/Module.zig
@@ -4865,14 +4865,31 @@ pub fn importFile(
     };
 }
 
-pub fn embedFile(mod: *Module, cur_file: *File, rel_file_path: []const u8) !*EmbedFile {
+pub fn embedFile(mod: *Module, cur_file: *File, import_string: []const u8) !*EmbedFile {
     const gpa = mod.gpa;
 
-    // The resolved path is used as the key in the table, to detect if
-    // a file refers to the same as another, despite different relative paths.
+    if (cur_file.pkg.table.get(import_string)) |pkg| {
+        const resolved_path = try std.fs.path.resolve(gpa, &[_][]const u8{
+            pkg.root_src_directory.path orelse ".", pkg.root_src_path,
+        });
+        var keep_resolved_path = false;
+        defer if (!keep_resolved_path) gpa.free(resolved_path);
+
+        const gop = try mod.embed_table.getOrPut(gpa, resolved_path);
+        errdefer assert(mod.embed_table.remove(resolved_path));
+        if (gop.found_existing) return gop.value_ptr.*;
+
+        const sub_file_path = try gpa.dupe(u8, pkg.root_src_path);
+        errdefer gpa.free(sub_file_path);
+
+        return newEmbedFile(mod, pkg, sub_file_path, resolved_path, &keep_resolved_path, gop);
+    }
+
+    // The resolved path is used as the key in the table, to detect if a file
+    // refers to the same as another, despite different relative paths.
     const cur_pkg_dir_path = cur_file.pkg.root_src_directory.path orelse ".";
     const resolved_path = try std.fs.path.resolve(gpa, &[_][]const u8{
-        cur_pkg_dir_path, cur_file.sub_file_path, "..", rel_file_path,
+        cur_pkg_dir_path, cur_file.sub_file_path, "..", import_string,
     });
     var keep_resolved_path = false;
     defer if (!keep_resolved_path) gpa.free(resolved_path);
@@ -4880,9 +4897,6 @@ pub fn embedFile(mod: *Module, cur_file: *File, rel_file_path: []const u8) !*Emb
     const gop = try mod.embed_table.getOrPut(gpa, resolved_path);
     errdefer assert(mod.embed_table.remove(resolved_path));
     if (gop.found_existing) return gop.value_ptr.*;
-
-    const new_file = try gpa.create(EmbedFile);
-    errdefer gpa.destroy(new_file);
 
     const resolved_root_path = try std.fs.path.resolve(gpa, &[_][]const u8{cur_pkg_dir_path});
     defer gpa.free(resolved_root_path);
@@ -4902,7 +4916,23 @@ pub fn embedFile(mod: *Module, cur_file: *File, rel_file_path: []const u8) !*Emb
     };
     errdefer gpa.free(sub_file_path);
 
-    var file = try cur_file.pkg.root_src_directory.handle.openFile(sub_file_path, .{});
+    return newEmbedFile(mod, cur_file.pkg, sub_file_path, resolved_path, &keep_resolved_path, gop);
+}
+
+fn newEmbedFile(
+    mod: *Module,
+    pkg: *Package,
+    sub_file_path: []const u8,
+    resolved_path: []const u8,
+    keep_resolved_path: *bool,
+    gop: std.StringHashMapUnmanaged(*EmbedFile).GetOrPutResult,
+) !*EmbedFile {
+    const gpa = mod.gpa;
+
+    const new_file = try gpa.create(EmbedFile);
+    errdefer gpa.destroy(new_file);
+
+    var file = try pkg.root_src_directory.handle.openFile(sub_file_path, .{});
     defer file.close();
 
     const actual_stat = try file.stat();
@@ -4915,10 +4945,6 @@ pub fn embedFile(mod: *Module, cur_file: *File, rel_file_path: []const u8) !*Emb
     const bytes = try file.readToEndAllocOptions(gpa, std.math.maxInt(u32), size_usize, 1, 0);
     errdefer gpa.free(bytes);
 
-    log.debug("new embedFile. resolved_root_path={s}, resolved_path={s}, sub_file_path={s}, rel_file_path={s}", .{
-        resolved_root_path, resolved_path, sub_file_path, rel_file_path,
-    });
-
     if (mod.comp.whole_cache_manifest) |whole_cache_manifest| {
         const copied_resolved_path = try gpa.dupe(u8, resolved_path);
         errdefer gpa.free(copied_resolved_path);
@@ -4927,13 +4953,13 @@ pub fn embedFile(mod: *Module, cur_file: *File, rel_file_path: []const u8) !*Emb
         try whole_cache_manifest.addFilePostContents(copied_resolved_path, bytes, stat);
     }
 
-    keep_resolved_path = true; // It's now owned by embed_table.
+    keep_resolved_path.* = true; // It's now owned by embed_table.
     gop.value_ptr.* = new_file;
     new_file.* = .{
         .sub_file_path = sub_file_path,
         .bytes = bytes,
         .stat = stat,
-        .pkg = cur_file.pkg,
+        .pkg = pkg,
         .owner_decl = undefined, // Set by Sema immediately after this function returns.
     };
     return new_file;

--- a/test/standalone.zig
+++ b/test/standalone.zig
@@ -102,4 +102,5 @@ pub fn addCases(cases: *tests.StandaloneContext) void {
     cases.addBuildFile("test/standalone/issue_13030/build.zig", .{ .build_modes = true });
     cases.addBuildFile("test/standalone/emit_asm_and_bin/build.zig", .{});
     cases.addBuildFile("test/standalone/issue_12588/build.zig", .{});
+    cases.addBuildFile("test/standalone/embed_generated_file/build.zig", .{});
 }

--- a/test/standalone/embed_generated_file/bootloader.zig
+++ b/test/standalone/embed_generated_file/bootloader.zig
@@ -1,0 +1,1 @@
+pub export fn _start() void {}

--- a/test/standalone/embed_generated_file/build.zig
+++ b/test/standalone/embed_generated_file/build.zig
@@ -1,0 +1,28 @@
+const std = @import("std");
+
+pub fn build(b: *std.Build) void {
+    const target = b.standardTargetOptions(.{});
+    const optimize = b.standardOptimizeOption(.{});
+
+    const bootloader = b.addExecutable(.{
+        .name = "bootloader",
+        .root_source_file = .{ .path = "bootloader.zig" },
+        .target = .{
+            .cpu_arch = .x86,
+            .os_tag = .freestanding,
+        },
+        .optimize = .ReleaseSmall,
+    });
+
+    const exe = b.addTest(.{
+        .root_source_file = .{ .path = "main.zig" },
+        .target = target,
+        .optimize = optimize,
+    });
+    exe.addAnonymousModule("bootloader.elf", .{
+        .source_file = bootloader.getOutputSource(),
+    });
+
+    const test_step = b.step("test", "Test the program");
+    test_step.dependOn(&exe.step);
+}

--- a/test/standalone/embed_generated_file/main.zig
+++ b/test/standalone/embed_generated_file/main.zig
@@ -1,0 +1,8 @@
+const std = @import("std");
+const blah = @embedFile("bootloader.elf");
+
+test {
+    comptime {
+        std.debug.assert(std.mem.eql(u8, blah[1..][0..3], "ELF"));
+    }
+}

